consensus/cryptoiface.go

```diff
@@ -0,0 +1,82 @@
+package consensus
+
+import (
+	"github.com/relab/hotstuff"
+)
+
+// CryptoBase provides the basic cryptographic methods needed to create, verify, and combine signatures.
+type CryptoBase interface {
+	// Sign creates a cryptographic signature of the given message.
+	Sign(message []byte) (signature QuorumSignature, err error)
+	// Verify verifies the given cryptographic signature according to the specified options.
+	// NOTE: One of either VerifySingle or VerifyMulti options MUST be specified,
+	// otherwise this function will have nothing to verify the signature against.
+	Verify(signature QuorumSignature, options ...VerifyOption) bool
+	// Combine combines multiple signatures into a single signature.
+	Combine(signatures ...QuorumSignature) (signature QuorumSignature, err error)
+}
+
+// VerifyOption sets options for the Verify function in the CryptoBase interface.
+type VerifyOption func(o *VerifyOptions)
+
+// VerifyThreshold specifies the number of signers that must have participated
+// in the signing of the signature in order for Verify to return true.
+func VerifyThreshold(threshold int) VerifyOption {
+	return func(o *VerifyOptions) {
+		o.Threshold = threshold
+	}
+}
+
+// VerifySingle verifies the signature against the given message.
+//
+// NOTE: VerifySingle and VerifyMulti cannot be used together,
+// and whichever option is specified last will be the effective one.
+func VerifySingle(message []byte) VerifyOption {
+	return func(o *VerifyOptions) {
+		o.MultipleMessages = false
+		o.Message = &message
+	}
+}
+
+// VerifyMulti verifies the signature against the multiple messages present in the given map.
+//
+// NOTE: VerifySingle and VerifyMulti cannot be used together,
+// and whichever option is specified last will be the effective one.
+func VerifyMulti(messages map[hotstuff.ID][]byte) VerifyOption {
+	return func(o *VerifyOptions) {
+		o.MultipleMessages = true
+		o.MessageMap = messages
+	}
+}
+
+// VerifyOptions specify options for the Verify function in the CryptoBase interface.
+type VerifyOptions struct {
+	// using a pointer for the Message such that we can tell if it is unspecified
+
+	Message          *[]byte
+	MessageMap       map[hotstuff.ID][]byte
+	MultipleMessages bool // true if `MessageMap` should be used instead of `Message`
+	Threshold        int
+}
+
+// Crypto implements the methods required to create and verify signatures and certificates.
+// This is a higher level interface that is implemented by the crypto package itself.
+type Crypto interface {
```

Comment:
I'm not a fan of big interfaces like this. Any reason to have all these methods in the same interface?

If you make separate interfaces for each method or pairs of methods, it would be easier to implement alternatives. Although, I'm not sure we will have multiple implementations of these methods.

Reply:
The reason we have this interface is to avoid an import cycle. The only implementor is the `crypto` struct in `crypto/crypto.go`, and I don't see any reason for alternative implementations. But the methods of the `crypto` struct are needed by the other modules. So we can have this interface, or we would have to move the whole `crypto` struct into the `consensus` package as well. Maybe there is a better way, but I haven't been able to think of one yet.